controllers/ingressnodefirewall_controller.go

```diff
@@ -56,96 +56,138 @@ type IngressNodeFirewallReconciler struct {
 	ingressNodeFirewallList := &infv1alpha1.IngressNodeFirewallList{}
 	if err := r.List(ctx, ingressNodeFirewallList); err != nil {
 		r.Log.Error(err, "Failed to list IngressNodeFirewall objects")
 		return ctrl.Result{}, err
 	}
-	nodeStateDesiredSpecs, err := r.buildNodeStateSpecs(ctx, ingressNodeFirewallList)
+	r.Log.Info("Building the desired node state specs", "req.Name", req.Name)
+	desiredNodeStates, err := r.buildNodeStates(ctx, ingressNodeFirewallList)
 	if err != nil {
 		r.Log.Error(err, "Failed to build IngressNodeFirewallNodeState")
 		return ctrl.Result{}, err
 	}
 
-	// 2) Get all IngressNodeFirewallNodeState objects in namespace <namespace> (namespace scoped).
-	ingressNodeFirewallNodeStateList := infv1alpha1.IngressNodeFirewallNodeStateList{}
-	listOpts := []client.ListOption{
-		client.InNamespace(r.Namespace),
-	}
-	if err := r.List(ctx, &ingressNodeFirewallNodeStateList, listOpts...); err != nil {
-		r.Log.Error(err, "Failed to list ingress node firewalls")
-		return ctrl.Result{}, err
-	}
-
 	// 3) Delete objects which should not be there, update existing objects and create missing ones.
 	var name string
 	for _, ingressNodeFirewallCurrentNodeState := range ingressNodeFirewallNodeStateList.Items {
 		// First, check if the desired specs contain a possibly new spec for the current node state object.
 		name = ingressNodeFirewallCurrentNodeState.Name
-		nodeStateDesiredSpec, ok := nodeStateDesiredSpecs[name]
+		desiredNodeState, ok := desiredNodeStates[name]
 		// If the current state object is not found in the list of desired states, then we must delete the current node
 		// state object.
 		if !ok {
+			r.Log.Info("Existing object not found in desired list, triggering deletion", "req.Name", req.Name)
 			if err := r.Delete(ctx, &ingressNodeFirewallCurrentNodeState); err != nil && !errors.IsNotFound(err) {
 				r.Log.Error(err, "Failed to delete node state",
+					"req.Name", req.Name,
 					"ingressNodeFirewallNodeState.Namespace", ingressNodeFirewallCurrentNodeState.Namespace,
 					"ingressNodeFirewallNodeState.Name", ingressNodeFirewallCurrentNodeState.Name)
 				return ctrl.Result{}, err
 			}
-			r.Log.Info("Deleted object",
-				"ingressNodeFirewallNodeState.Namespace", ingressNodeFirewallCurrentNodeState.Namespace,
-				"ingressNodeFirewallNodeState.Name", ingressNodeFirewallCurrentNodeState.Name)
 			continue
 		}
 
-		// If the node name was found, but the specs are equal, continue with the next current node state.
-		// Also, remove the object from the nodeStateDesiredSpecs so that we can later iterate over the items
-		// that must still be created.
-		if equality.Semantic.DeepEqual(ingressNodeFirewallCurrentNodeState.Spec, nodeStateDesiredSpec) {
-			r.Log.Info("No change detected, skipping",
+		// If the node name was found ...
+		// a) compare the specs. If the specs are different, update the current spec.
+		if !equality.Semantic.DeepEqual(ingressNodeFirewallCurrentNodeState.Spec, desiredNodeState.Spec) {
```

Comment:
can u have a generic merge lib and use it to do the common merge parts and u can pass func in case obj assign spec for status update the status ?

Reply:
Let's talk about that tomorrow